DM-45872: Make the new query system public

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

TallJimbo committed:
Handle empty query 'where' strings.

parse_expression can in fact return, as I've discovered from
experience, contrary to its type annotations.

By considering empty strings to evaluate to True, they'll get
simplified away, which is consistent with the where string not being
provided and a convenience to at least QG generation (at it's easier
to implement than raising an exception).

TallJimbo committed:
Fix copy-paste bug in expression factory dataset timespan access.

Aggressive error handling + __getattr__ was masking this.

timj committed:
Rename Butler._query to Butler.query

timj committed:
Make the ButlerCollections.query methods public

timj committed:
Make Butler.query_datasets and others public

No tests yet.

timj committed:
Remove the simple query APIs from hybrid butler

They are not needed because the default Butler implementation
works fine with remote butler.

timj committed:
Add some simple query interface tests

These are the advanced tests with the simple interface where
possible.

timj committed:
Add --limit to query-datasets and related command-line scripts

Also removed some duplicate code in retrieve-artifacts and
transfer-from.

timj committed:
Add --order-by support to query-datasets command-line

Do not include it in butler associate.
This required a small rewrite of the table accumulator to use
a dict rather than a set.

timj committed:
Add news fragment

timj committed:
Add a test for query-datasets command-line with limit/order_by

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1584,7 +1588,7 @@ def query_datasets(
         collections : collection expression, optional
             A collection name or iterable of collection names to search. If not
-            provided, the default collections are used.  See
+            provided, the default collections are used. Can be a wildcard. See
```
Since `find_first=True` is now the default, might want to add that this can be a wildcard only if `find_first=False` is passed as well.

timj replied:
Although as written it doesn't do that because I expand the collections before passing to the query system. I need to take another look. Thanks for the reminder.

timj replied:
I just tried it without me expanding the collections and the new query system doesn't look like it handles wildcards at all. I'm going to have to query the collections I'm given and then check if the query returns something different to what I started with and then complain if find_first.

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1573,6 +1575,8 @@ def _query_datasets(
         bind: Mapping[str, Any] | None = None,
         with_dimension_records: bool = False,
+        order_by: Iterable[str] | str | None = None,
+        limit: int = 20_000,
```
The default here should probably be the same as the CLI default, and I think it'd probably be best to make it negative and warn here, too, if a negative limit is hit.  As it is it's too easy to hit the limit and not know it.

timj replied:
Ok. I can move the negative limit support into the wrappers and out of the script (and also use -20_000 in both places). At some point we can have the conversation as to whether the actual limit support in the advanced query system should understand negative limit.

TallJimbo commented on python/lsst/daf/butler/script/queryDatasets.py:
```diff
@@ -46,21 +45,15 @@
     """
 
-    datasetRefs: set[_RefInfo]
+    datasetRefs: dict[DatasetRef, str | None]
```
Is the intent here to drop rows where a ref is associated with more than one URI?  I think that's the effect of this change.  If that's undesirable I think we could safely switch to `list[_RefInfo]` instead of `set[_RefInfo]` now.

timj replied:
Good point. I thought it was for handling the deduplication of results (and I now need to retain the returned sorting order) but it also needs to support disassembled composites so I need to check that by querying pipelines_check.

timj replied:
It was working already because the dataset type name included the component and there was a table per component. I've rewritten the other code that calls this to generate a table as data arrive rather than accumulating in memory and now `butler query-datasets` starts sending output as soon as the first dataset type results arrives.

TallJimbo replied:
What about multiple refs from chained datastores?

timj replied:
Ok. I have no means of testing this but I do have a fix that I made that stored lists of URIs that I can resurrect (I deleted it when I convinced myself this was only disassembled composites).

timj replied:
Ok. I  took a deeper look at this:

* `butler.get_many_uris` only returns one URI per ref so there can't be multiple URIs in this table.
* There is a chaining test that was working by luck and assuming the second datastore would return results. This was luck because there was a bug in the test in that the in-memory datastore was not being used because the butler instance got remade -- fixing that bug causes in-memory dataset URIs to be reported.
* I need to fix get_many_uris so that it won't return ephemeral URIs if permanent ones exist.
* I probably can remove my patch to support multiple URIs since it won't happen.

TallJimbo commented on python/lsst/daf/butler/script/queryDatasets.py:
```diff
@@ -111,26 +104,29 @@ def getAstropyTable(self, datasetTypeName: str) -> AstropyTable:
         dataset_table = AstropyTable(np.array(rows), names=columnNames, dtype=columnTypes)
-        return sortAstropyTable(dataset_table, dimensions, ["type", "run"])
+        if sort:
+            return sortAstropyTable(dataset_table, dimensions, ["type", "run"])
```
I had no idea we had a separate sorting code path for `order_by` on the CLI that did it all in memory, and I'm not thrilled by that, in terms of the maintenance load and keeping behavior consistent across the two code paths so we have the freedom to change the implementation behind the scenes (it already looks inconsistent in at least not supporting descending sorts with `-identifier` syntax).  I also have no idea why it's trying to sort by spatial columns (what does that even mean?) and the list of dimensions it takes is not the right way to work out spatial/temporal things (which may be dimension elements, not dimensions).

I do think it might make sense to do some in-memory sorting to make output order deterministic when `order_by` is not given, but that's a lot simpler that reimplementing `order_by` in Python.

Fixing all that in the existing CLI commands is out of scope here, but I think I'd prefer we not add new functionality that depends on it here.  For now, could we support `--order-by` in `query-datasets` only on single-dataset-type queries (which we can do in the DB), and revisit it for multiple-dataset-type queries when we have the new Python methods for that?

timj replied:
It's only ever sorting each table by dataset type and not trying to sort across dataset types. Nate P was trying to make it so that the results were returned in a consistent way each time and it didn't seem to be a huge issue given people shouldn't be querying for millions of rows on the command line.  If the user specifies `--order-by` no sorting happens here so I think that's what you are asking for and if someone asks to order by with a dimension that is not part of the dataset type the query breaks before any of this code is hit.

TallJimbo commented on python/lsst/daf/butler/script/queryDatasets.py:
```diff
@@ -225,30 +239,59 @@ def getDatasets(self) -> Iterator[DatasetRef]:
+                    # the list.
+                    results.pop(-1)
+            _LOG.debug("Got %d results for dataset type %s", len(results), dt)
+            yield from results
```
Are we actually getting anything out of lazy iteration here, or do all callers eventually just put the results into regular in-memory container anyway?  For at least the case where we turn it into an Astropy table for formatting, I'm sure it'd be better to pass a `list[DatasetRef]` so Astropy can know the number of rows in advance and allocate up front accordingly.

timj replied:
Yielding seemed nice given the looping over dataset types meaning that we don't require we keep everything in memory. I have been wondering why the command line doesn't try to dump the tables for each dataset type as they arrive from the system rather than waiting until the queries complete. If it yielded `list[DatasetRef]` for each dataset type then that would be more flexible.

timj replied:
I made this change to yield the results of each dataset type and then yield each table immediately. Much more responsive.

TallJimbo commented on python/lsst/daf/butler/script/queryDatasets.py:
```diff
@@ -225,30 +239,59 @@ def getDatasets(self) -> Iterator[DatasetRef]:
+                        "Requested limit of %d hit for number of datasets returned. "
+                        "Use --limit to increase this limit.",
+                        self._limit,
+                    )
```
I think we should move this warn-if-limit-hit logic into the Python API, even if we have to do it here too (at least until we implement multiple dataset type queries as a new Python interface).  That could be another ticket, of course.

timj replied:
Ok. I can remove this warning completely here if the python API is going to handle it.

timj replied:
I added the warning into `Butler.query_datasets` but had to leave the warning here because the context is different. If you ask for 1000 results spread over 20 dataset types you want to know the limit is 1000 but the internal query only thinks the limit was 100 or whatever.

timj committed:
Be more explicit about the default status of --find-first

timj committed:
Fix handling of find_first=True with collection wildcards

timj committed:
Add support for negative limit to Butler.query_datasets

timj committed:
Use same limit for command line query-datasets and butler.query_datasets

timj committed:
Change command-line to convert limit=0 to limit=None for API usage

timj committed:
Fix --show-uri for chained datasstores

The ref will be the same so we need to store multiple
URIs for that ref.

timj committed:
Change query-datasets command line to use iterators

Now output begins appearing as soon as the first dataset type
has been queried. No longer accumulates all the results into
memory before writing them and so feels more responsive with
lower memory footprint.

Does not paginate queries for a single dataset type.

timj committed:
Add support for negative limit to query_dimension_records

timj committed:
Add negative limit support to query_data_ids

timj committed:
Make sure query-datasets complains if collection wildcard with find-first

timj committed:
Rename Query.general to x_general to mark as experimental

timj committed:
Make in-memory dataset URI more informative and remove duplication

timj committed:
Set large but negative limits for query_data_ids and query_dimension_records

timj committed:
Only use ephemeral URIs when we need them in get_many_uris

timj committed:
Change query-datasets command-line to use the butler it creates

Without this change the chained datastore created and populated
is not used in the actual chaining test and so the tests are
not testing what it is stated to test.

timj committed:
Fix mypy validation error by adding explicit type guard function

Fixes:

python/lsst/daf/butler/queries/expression_factory.py:468: error: Argument "field" to "DatasetFieldReference" has incompatible type "str";
expected "Literal['dataset_id', 'ingest_date', 'run', 'collection', 'timespan']"  [arg-type]

triggered by pydantic 2.9.0.

timj committed:
Allow MetricTestRepo to use a different storage class

This allows tests to run with and without composite disassembly.

timj committed:
Add a test for query-datasets uris without disassembly

timj committed:
Revert "Fix --show-uri for chained datasstores"

This reverts commit a287d0bd7ebe3a989f298acda666d07442d735da.

After some thought it became clear that get_many_uris guarantees
only a single URI per dataset ref regardless of whether a
chained datastore is present so this additional complexity is
not needed. Tables are created for each component dataset type
so even for composite disassembly there will not be multiple
URIs in a table for the same dataset ref.
